src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOptimizedNullableOperators.cs

```diff
@@ -3486,5 +3536,208 @@ .maxstack  2
             verifier.VerifyDiagnostics();
             verifier.VerifyIL("C.M", il);
         }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/52629")]
+        public void NullableConstant_Decimal_E0()
+        {
+            var source = """
+                C.Run(0m);
+                C.Run(0.0m);
+                C.Run(1m);
+                C.Run(2m);
+                C.Run(null);
+
+                class C
+                {
+                    public static void Run(decimal? x)
+                    {
+                        System.Console.Write(M(x) ? 1 : 0);
+                    }
+                    static bool M(decimal? x) => x == 0m;
+                }
+                """;
+            var output = "11000";
+            var il = """
+                {
+                  // Code size       32 (0x20)
+                  .maxstack  2
+                  .locals init (decimal? V_0,
+                                decimal V_1)
+                  IL_0000:  ldarg.0
+                  IL_0001:  stloc.0
+                  IL_0002:  ldloca.s   V_1
+                  IL_0004:  initobj    "decimal"
+                  IL_000a:  ldloca.s   V_0
+                  IL_000c:  call       "decimal decimal?.GetValueOrDefault()"
+                  IL_0011:  ldloc.1
+                  IL_0012:  call       "bool decimal.op_Equality(decimal, decimal)"
+                  IL_0017:  ldloca.s   V_0
+                  IL_0019:  call       "bool decimal?.HasValue.get"
+                  IL_001e:  and
+                  IL_001f:  ret
+                }
+                """;
+            var verifier = CompileAndVerify(source, options: TestOptions.ReleaseExe, expectedOutput: output);
+            verifier.VerifyDiagnostics();
+            verifier.VerifyIL("C.M", il);
+            verifier = CompileAndVerify(source, options: TestOptions.DebugExe, expectedOutput: output);
+            verifier.VerifyDiagnostics();
+            verifier.VerifyIL("C.M", il);
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/52629")]
+        [WorkItem("https://github.com/dotnet/roslyn/issues/73510")]
+        public void NullableConstant_Decimal_E00()
+        {
+            var source = """
+                C.Run(0m);
+                C.Run(0.0m);
+                C.Run(1m);
+                C.Run(2m);
+                C.Run(null);
+
+                class C
+                {
+                    public static void Run(decimal? x)
+                    {
+                        System.Console.Write(M(x) ? 1 : 0);
+                    }
+                    static bool M(decimal? x) => x == 0.0m;
```

Comment:
>x == 0.0m; [](http://example.com/codeflow?start=49&length=10)

Are we also testing `0.0m == x`?

Reply:
Consider testing `x == 0.00m` and `0.00m == x` since the reported issue uses `0.00m`.

Reply:
Let's also test `x != <constant>` and `<constant> != x`.